backend/src/main/java/com/google/sps/FindMatchQuery.java

```diff
@@ -50,30 +54,77 @@ public Match findMatch(Participant newParticipant) {
     List<Participant> sameDurationParticipants =
         participantDatastore.getParticipantsWithDuration(duration);
 
-    // Set reference date time using clock
-    long currentTimeMillis = clock.millis();
-
     long newEndTimeAvailable = newParticipant.getEndTimeAvailable();
+    MatchPreference newMatchPreference = newParticipant.getMatchPreference();
+    String newRole = newParticipant.getRole();
+    String newProductArea = newParticipant.getProductArea();
 
     // Compare new participant preferences with other participants to find match
     for (Participant currParticipant : sameDurationParticipants) {
       // Check if participants are both free for that duration + extra
       long currEndTimeAvailable = currParticipant.getEndTimeAvailable();
-      long earliestEndTimeAvailable = Math.min(newEndTimeAvailable, currEndTimeAvailable);
-      boolean compatibleTime =
-          (currentTimeMillis + TimeUnit.MINUTES.toMillis(duration + PADDING_MINUTES))
-              < earliestEndTimeAvailable;
+      if (!isCompatibleTime(duration, newEndTimeAvailable, currEndTimeAvailable)) {
+        System.out.println("not compatible time");
+        continue;
+      }
+
+      // Check match preference compatibility and get combined preference if compatible
+      MatchPreference currMatchPreference = currParticipant.getMatchPreference();
+      if (!isCompatibleMatchPreference(newMatchPreference, currMatchPreference)) {
+        System.out.println("not compatible match pref");
+        continue;
+      }
+      MatchPreference combinedMatchPreference =
+          getCombinedMatchPreference(newMatchPreference, currMatchPreference);
 
-      if (compatibleTime) {
-        // Create and return match
-        return new Match(
-            newParticipant.getUsername(),
-            currParticipant.getUsername(),
-            duration,
-            currentTimeMillis);
+      if (combinedMatchPreference != MatchPreference.ANY) {
+        int numSameFields = getNumSameFields(newParticipant, currParticipant);
+        System.out.println(numSameFields);
+        if (combinedMatchPreference == MatchPreference.SIMILAR && numSameFields < MIN_SAME_FIELDS) {
+          continue;
+        } else if (combinedMatchPreference == MatchPreference.DIFFERENT
+            && numSameFields >= MIN_SAME_FIELDS) {
+          continue;
+        }
```

Comment:
To be consistent with `isCompatibleTime ` and `isCompatibleMatchPreference`, can we encapsulate this part into `isMatchPreferenceSatisfied(newMatchPreference, currMatchPreference)`?